graphql_compiler/cost_estimation/filter_selectivity_utils.py

```diff
@@ -44,6 +51,138 @@ def _are_filter_fields_uniquely_indexed(filter_fields, unique_indexes):
+        query_interval[0] = max(query_interval[0], parameter_values[0])
+    elif filter_operator == '<':
+        # This (exclusively) constrains the upper bound of the values passing through the filter.
+        query_interval[1] = min(query_interval[1], parameter_values[0] - 1)
+    elif filter_operator == '<=':
+        # This (inclusively) constrains the upper bound of the values passing through the filter.
+        query_interval[1] = min(query_interval[1], parameter_values[0])
+    elif filter_operator == 'between':
+        # This (inclusively) constrains the lower and upper bounds of the
+        # values passing through the filter.
+        query_interval[0] = max(query_interval[0], parameter_values[0])
+        query_interval[1] = min(query_interval[1], parameter_values[1])
+    else:
+        raise AssertionError(u'Cost estimator found unsupported integer inequality operator {}'
+                             .format(filter_operator))
+
+    # If the interval of values passing through the filters is empty, no results will be
+    # returned. This happens if the interval's upper bound is smaller than the lower bound.
+    if query_interval[0] > query_interval[1]:
+        field_selectivity = Selectivity(kind=ABSOLUTE_SELECTIVITY, value=0.0)
+        return field_selectivity
+
+    if not domain_interval[0] <= query_interval[0] <= query_interval[1] <= domain_interval[1]:
+        raise AssertionError(u'Query interval {} is not'
+                             u'a subset of the given domain interval {}.'
+                             .format(query_interval, domain_interval))
+
+    # Assumption: the values of the integers being filtered are evenly distributed among the domain
+    # of valid values.
+    query_interval_size = query_interval[1] - query_interval[0] + 1
+    domain_interval_size = domain_interval[1] - domain_interval[0] + 1
+    fraction_of_domain_queried = query_interval_size / domain_interval_size
+
+    field_selectivity = Selectivity(
+        kind=FRACTIONAL_SELECTIVITY,
+        value=query_interval_size / domain_interval_size
+    )
+    return field_selectivity
+
+
+def _estimate_inequality_filter_selectivity(
+    schema_graph, statistics, filter_info, parameters, location_name
+):
+    """Calculate the selectivity of a specific inequality filter at a given location.
+
+    Args:
+        schema_graph: SchemaGraph object
+        statistics: Statistics object
+        filter_info: FilterInfo object, filter on the location being filtered
+        parameters: dict, parameters with which query will be executed
+        location_name: string, type of the location being filtered
+
+    Returns:
+        Selectivity object, the selectivity of a specific inequality filter at a given location.
+    """
+    filter_operator = filter_info.op_name
+    if filter_operator not in INEQUALITY_OPERATORS:
+        raise ValueError(u'Inequality filter selectivity estimator received a filter'
```

Comment:
`filter` and `with` don't have any whitespace between them, so this will end up saying `received a filterwith non-inequality`